### DISCUSSION TITLE
Why doesn't count() accept include?

### DISCUSSION THREAD

**Discussion by djdmbrwsk** [prisma/prisma#16148] [Q&A]
Until is https://github.com/prisma/prisma/issues/7550 implemented, as recommended in https://github.com/prisma/prisma/discussions/8830#discussion-3529675, I'm trying to make 2 queries to get a page of records as well as the total record count. However, since I'm filtering by a relationship's field I'm unable to pass the same `where` / `include` query into  both `findMany()` and `count()`.

```ts
const where = {
  user: {
    name: 'foobar',
  },
};

const include = {
  user: {
    select: {
      name: true,
    },
  },
};

const [posts, totalCount] = await Promise.all([
  prisma.post.findMany({
    where,
    include,
    skip: 0,
    take: 50,
  }),
  prisma.post.count({
    where,
    include, // This is not allowed
  }),
]);
```

I can hack another `findMany()` to make this work, but that seems horribly inefficient. Shouldn't the aim be to make `count()` and `findMany()` have the same signature except for the return type?

Anyone have any guidance? Thanks!

**Comment by nurul3101** [comment]
Hey @djdmbrwsk 👋 
As you mentioned include is not supported in the count yet. 
Did you have a look at `filteredRelationCount` preview feature? [Reference](https://www.prisma.io/docs/concepts/components/prisma-client/aggregation-grouping-summarizing#filter-the-relation-count). 
I think this could be helpful to you.

**Reply by djdmbrwsk** [reply]
Hey there! I did see that feature, but it looks like it’s for getting the count of child records where the relation is one-to-many?

In my case I’m looking at get the count of records in the table I’m operating on while filtering by a relations field.

**Reply by pantharshit00** [reply]
@djdmbrwsk Can you please tell what exactly you are passing in `where` and `include` variables? Do you want `count` to have `include` which can count the rows of related fields?

**Reply by djdmbrwsk** [reply]
Given these records in a DB:
```
// Users
{ id: 1, name: 'foobar' }
{ id: 2, name: 'barbaz' }

// Posts
{ id: 1, title: 'Post One', userId: 1 }
{ id: 2, title: 'Post Two', userId: 1 }
{ id: 3, title: 'Post Three', userId: 2 }
```

I want to call `count()` like this:
```
prisma.post.count({
  where: {
    user: {
      name: 'foobar',
    },
  },
  include: {
    user: {
      select: {
        name: true,
      },
    },
  }
}),
```

And I'm expecting the return value `2`.

My problem is that the `include` field is not allowed when calling `count()`.

**Reply by nurul3101** [reply]
@djdmbrwsk 👋 
Considering this schema:
```prisma
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["filteredRelationCount", "extendedWhereUnique"]
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model Users {
  id        Int      @id @default(autoincrement())
  email     String   @unique
  password  String
  name      String
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  posts     Posts[]
}

model Posts {
  id        Int     @id @default(autoincrement())
  title     String
  content   String
  published Boolean @default(false)
  author    Users?  @relation(fields: [authorId], references: [id])
  authorId  Int?
}
```

And this script:
```typescript
import { PrismaClient } from '@prisma/client';

const prisma = new PrismaClient({
  log: ['query', 'info', 'warn'],
});

async function main() {
  await prisma.users.create({
    data: {
      id: 1,
      email: 'adam@prisma.io',
      name: 'Adam',
      password: '1234',
      posts: {
        createMany: {
          data: [
            {
              content: 'Hello World',
              title: 'Hello World',
            },
            {
              content: 'Hello World 2',
              title: 'Hello World 2',
            },
          ],
        },
      },
    },
  });

  await prisma.users.create({
    data: {
      id: 2,
      email: 'beta@prisma.io',
      name: 'Beta',
      password: '1234',
      posts: {
        createMany: {
          data: [
            {
              content: 'Hello World 3',
              title: 'Hello World 3',
            },
          ],
        },
      },
    },
  });

  const records = await prisma.users.findMany({
    where: {
      email: 'adam@prisma.io',
    },
    include: {
      _count: {
        select: {
          posts: true,
        },
      },
    },
  });

  console.log(records);
}

main()
  .catch((e) => {
    throw e;
  })
  .finally(async () => {
    await prisma.$disconnect();
  });

```

Response:
```
[
  {
    id: 1,
    email: 'adam@prisma.io',
    password: '1234',
    name: 'Adam',
    createdAt: 2022-11-17T12:16:44.251Z,
    updatedAt: 2022-11-17T12:16:44.251Z,
    _count: { posts: 2 }
  }
]
```

Would this solution of using _count with include work for you?

**Reply by djdmbrwsk** [reply]
Maybe! I need to be able to filter parent records AND child records at the same time, getting the count of the children. After a quick Google, it [seems possible](https://www.prisma.io/docs/concepts/components/prisma-client/aggregation-grouping-summarizing#filter-the-relation-count). Will give it a shot and report back.

**Reply by nurul3101** [reply]
Sure, do let us know in case you run into any issues. 😄

**Reply by anitacung** [reply]
👋🏻 Hi @djdmbrwsk, 

I'm going through pending Github Discussions to check if you need any help. Did Nurul’s solution help solve your issue?

If there's anything we can help you with, I'd be happy to organize a call with our Product Team. 

Let me know! 🤗

**Reply by djdmbrwsk** [reply]
Finally had some time to play with this. While the `_count` approach does work to filter parent and child records at the same time and return a count, it's not exactly the count I'm interested in. I'm trying to get the **total** count of the child records. I guess I could sum the result of the `_count` query, but that's not very efficient. Additionally, it sure would be cleaner to use the same `where` object on both my paged `getMany()` call and my `count()` call.

```ts
// Vars
const where = {
  status: 'active',
  user: {
    name: 'foobar',
  },
};

const include = {
  user: {
    select: {
      name: true,
    },
  },
};

// _count version
const pagedPosts = await prisma.post.findMany({
  where,
  include,
  skip: 0,
  take: 50,
});
const usersWithPostCount = prisma.user.findMany({
  where: where.user,
  select: {
    _count: {
      select: {
        posts: where,
      },
    },
  },
});
const totalPosts = usersWithPostCount.reduce((t, i) => t + i, 0);

// Desired version
const pagedPosts = await prisma.post.findMany({
  where,
  include,
  skip: 0,
  take: 50,
});
const totalPosts = prisma.post.count({
  where,
  include, // This is not allowed
});
```

**Reply by nurul3101** [reply]
Do you mind adding this use case to the original feature request so that our engineering team can take care of it while working on the feature request?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
